### PR TITLE
Fix done button click not blur in iOS keyboard

### PR DESCRIPTION
**Pull request by JunbinDeng**
- https://github.com/flutter/flutter/issues/96357

And besides I found the following issue is related to TextField focus:

- https://github.com/flutter/flutter/issues/91755

This commit resolves iOS virtual keyboard glitch and TextField focus issue in Safari. When tapping the done button above the keyboard, TextField blurs and closes the keyboard right away instead of refocusing and reopening afterward.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Comment by justinmc** [comment]
CC @mdebbar

**Comment by JunbinDeng** [comment]
> Could you elaborate a little more about how this PR fixes the clicks on "done"?

Thanks for reviewing. In the beginning, I found that taping the 'done' button above the soft input in iOS Safari was not working correctly as the keyboard closed and opened up again. 

When pressing the 'done' button, the event listener 'onblur' should trigger a callback then the keyboard should be closed normally. The reason why the keyboard kept opening up was that the window was still on focus.

I was trying to remove everything except 'owner.sendTextConnectionClosedToFrameworkIfAny' in the 'onblur' listener callback. But a new problem occurred, which was tapping the TextInput could not bring up the keyboard occasionally.

I figured out the 'onblur' listener callback was triggered, as long as the listener had set up after a while. It should call by mistake.

Having tested several times, I found that the intervals between the 'onblur' event listener registered and callback triggered were below 200ms. 

It was considered a fast callback, which should regain focus instead of blur.

Please let me know if there's any other way I can do better.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart [hunk 1]
```diff
@@ -1465,7 +1478,7 @@ class IOSTextEditingStrategy extends GloballyPositionedTextEditingStrategy {
     //    okay because the virtual keyboard will hide, and as soon as the user
     //    taps the text field again, the virtual keyboard will come up.
     subscriptions.add(activeDomElement.onBlur.listen((_) {
-      if (windowHasFocus) {
+      if (DateTime.now().difference(blurSubscriptionStart) < _blurFastCallbackInterval) {
```
[the remark this answers is not in the document]
Comment (JunbinDeng) [reply lib/web_ui/lib/src/engine/text_editing/text_editing.dart:1481]: I suppose the `windowHasFocus` is always true in a fast callback case, so I left it out.

File: lib/web_ui/test/text_editing_test.dart [hunk 2]
```diff
@@ -675,6 +675,56 @@ void testMain() {
         // TODO(mdebbar): https://github.com/flutter/flutter/issues/50769
         skip: browserEngine == BrowserEngine.edge);
 
+    test('focus and disconnection with delaying blur in iOS', () async {
+      // Test on ios-safari only.
+      if (browserEngine == BrowserEngine.webkit &&
+          operatingSystem == OperatingSystem.iOs) {
```
[the remark this answers is not in the document]
Comment (JunbinDeng) [reply lib/web_ui/test/text_editing_test.dart]: I fixed it!

**Comment by mdebbar** [comment]
Thanks for the explanation!

I'm still trying to understand the "fast callback". Here's my understanding of the sequence of things:

The `onblur` event listener is registered when the text field is created (which happens when you tap on it for the first time). And the `onblur` event is triggered when you click the "done" button.

Why does the duration between these two things matter? And in what situation does the `onblur` trigger immediately after subscription?

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/text_editing/text_editing.dart:1481]: You can find an explanation of why we needed to do things the way they are now:
https://github.com/flutter/engine/blob/6cac347c3d39013ddc2bc0ad29894f01b9fc1aff/lib/web_ui/lib/src/engine/text_editing/text_editing.dart#L1456-L1466

I think you may be ignoring the case of minimizing the browser (or switching to another app). That's where `windowHasFocus` is useful.

**Comment by JunbinDeng** [comment]
> 

It could be more clear to deduce matters in this way.

The current situation:

1. Tap the TextField
2. Set up the blur listener
3. Trigger the blur listener callback immediately (should not happen)
4. Yes, the window has the focus
5. Keep showing the keyboard

The current keyboard action:

1. Press the 'done' button
2. The keyboard is hiding
3. Trigger the blur listener callback
4. Yes, the window has the focus
5. The keyboard shows up (unable to hide)

If we apply the 'fast callback' solution, then:

1. Tap the TextField
2. Set up the blur listener
3. Trigger the blur listener callback immediately (should not happen)
4. Yes, a 'fast callback'
5. Keep showing the keyboard

The solved keyboard action:

1. Press the 'done' button
2. The keyboard is hiding
3. Trigger the blur listener callback
4. No, not a 'fast callback'
5. The keyboard hides

Hopefully can answer your questions.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (JunbinDeng) [reply lib/web_ui/lib/src/engine/text_editing/text_editing.dart:1481]: That's right. Either minimizing the browser (or switching to another app) or tapping "done" puts an end to 'sendTextConnectionClosedToFrameworkIfAny'. However, the doc seems to be confusing without the `windowHasFocus`. I add it back.

**Review by mdebbar — APPROVED** [review]
Thanks for the great explanation! I added a suggestion on how to incorporate it in the doc comments. Feel free to re-phrase it if needed.

**Comment by JunbinDeng** [comment]
> Thanks for the great explanation! I added a suggestion on how to incorporate it in the doc comments. Feel free to re-phrase it if needed.

I appreciate the time and effort you have spent to share your insightful comments, which will be helpful to improve code readability.

**Comment by mdebbar** [comment]
Thanks for contributing this fix to the flutter web engine!

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart [hunk 3]
```diff
@@ -1389,6 +1389,16 @@ class IOSTextEditingStrategy extends GloballyPositionedTextEditingStrategy {
   Timer? _positionInputElementTimer;
   static const Duration _delayBeforePlacement = Duration(milliseconds: 100);
 
+  /// The interval is considered as fast between blur subscription and callback.
```
Comment (justinmc) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:1392]: Nit: Maybe this is more clear.

"
This interval between the blur subscription and callback is considered to be fast.
"

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart [hunk 4]
```diff
@@ -1389,6 +1389,16 @@ class IOSTextEditingStrategy extends GloballyPositionedTextEditingStrategy {
...
+  ///
+  /// This is only used for iOS. The blur callback may trigger as soon as
+  /// the creation of the subscription. In that case, the virtual keyboard will
+  /// show and hide again occasionally.
```
Comment (justinmc) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:1396]: Nit: Adding "quickly" to make it clear this is weird incorrect behavior.

"
Occasionally in that case, the virtual keyboard will quickly show and hide again.
"

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart [hunk 5]
```diff
@@ -1389,6 +1389,16 @@ class IOSTextEditingStrategy extends GloballyPositionedTextEditingStrategy {
...
+  ///
+  /// Lesser than the interval allows the virtual keyboard to keep showing up
```
Comment (justinmc) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:1398]: Nit: "Lesser than the" => "Less than this"

**Review by justinmc — APPROVED** [review]
LGTM 👍 

Just some small nits for my opinion on making the docs more readable, but it's up to you.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart [hunk 6]
```diff
@@ -1453,6 +1463,9 @@ class IOSTextEditingStrategy extends GloballyPositionedTextEditingStrategy {
 
     _addTapListener();
 
+    // Record start time of blur subscription.
+    final DateTime blurSubscriptionStart = DateTime.now();
```
Comment (kevmoo) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:1467]: Would be MUCH cheaper to use `StopWatch` – creating a `DateTime` object is a lot more expensive. We just need to track ticks!

Comment (kevmoo) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:1467]:
```suggestion
    final watch = StopWatch()..start();
```

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart [hunk 7]
```diff
@@ -1464,8 +1477,15 @@ class IOSTextEditingStrategy extends GloballyPositionedTextEditingStrategy {
...
     //    taps the text field again, the virtual keyboard will come up.
+    // 4. Safari sometimes sends a blur event immediately after activating the
+    //    input field. In this case, we want to keep the focus on the input field.
+    //    In order to detect this, we measure how much time has passed since the
+    //    input field was activated. If the time is too short, we re-focus the
+    //    input element.
     subscriptions.add(activeDomElement.onBlur.listen((_) {
-      if (windowHasFocus) {
+      final bool isFastCallback =
+          DateTime.now().difference(blurSubscriptionStart) < _blurFastCallbackInterval;
```
Comment (kevmoo) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:1487]: You're allocating ANOTHER `DateTime` here. With `Stopwatch` you can just call `watch.elapsed.difference`

**Comment by JunbinDeng** [comment]
> LGTM 👍
> 
> Just some small nits for my opinion on making the docs more readable, but it's up to you.

Thanks for your suggestion. I've done it.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart (hunk 6 above)
Comment (JunbinDeng) [reply lib/web_ui/lib/src/engine/text_editing/text_editing.dart]: Thanks for your suggestion. I've changed to use StopWatch.

**Pull request merged**